src/catch2/reporters/catch_reporter_listening.cpp

```diff
@@ -15,151 +17,157 @@ namespace Catch {
         m_listeners.push_back( std::move( listener ) );
     }
 
-    void ListeningReporter::addReporter(IStreamingReporterPtr&& reporter) {
-        assert(!m_reporter && "Listening reporter can wrap only 1 real reporter");
-        m_reporter = std::move( reporter );
-        m_preferences.shouldRedirectStdOut = m_reporter->getPreferences().shouldRedirectStdOut;
+    void ListeningReporter::addReporter( IStreamingReporterPtr&& reporter ) {
+        m_preferences.shouldRedirectStdOut |= reporter->getPreferences().shouldRedirectStdOut;
+
+        // We will need to output the captured stdout if there are reporters
+        // that do not want it captured.
+        m_haveNoncapturingListeners |= !reporter->getPreferences().shouldRedirectStdOut;
```

Comment:
I am actually not sure these are the correct semantics...

The basic idea behind capturing is that for reporters with structured output, we cannot let tests just print to stdout and break the format... However, do we care about this if the structured output is written elsewhere? Do we care about also printing the output if there is a reporter with non structured output? What if that reporter is redirected? 😕  I honestly don't know.

Reply:
> The basic idea behind capturing is that for reporters with structured output, we cannot let tests just print to stdout and break the format

I don't entirely agree with that. I think we want to capture the output so that we can include it in the structured output in some well-defined manner. I don't see any reason to support `./prints-to-stdout-test -r xml >out.xml`. There *is* a supported way to do this (`./prints-to-stdout-test -r xml:out.xml`, or even `-r xml -o out.xml`).

On the other hand, something like `./prints-to-stdout-test -r xml | process-the-xml` is not _that_ far-fetched anymore. In that case we really don't want the output to be reprinted to stdout.

However, it may be plausible to run `./test -r console -r xml:out.xml`, where I think I would want to see the output. OTOH in that case I can very well specify that explicitly as an opt-in.

Another conundrum is `./test -r console:out.txt -r xml:out.xml`. The stdout output on stdout will do us no good, unless of course it is unrelated to the tests at all and we want it separately. OTOH if we want it to be included in the txt report, it would mean we want the console reporter to *capture* the output and print it in a well-defined place. We could get this with a capturing console reporter, which would be trivial to implement.

**Summary**: Capturing reporters may want to include the captured output in some structured way. Non-capturing reporters do not care about stdio at all. Unless they are a capturing reporter in disguise and the structured way to display the output is to print it at a given place (at the end of testcase or at the place where it would be if the outputs were interleaved). But it never worked that way (`./test -r console -o out.txt`) and if we want that, we should get a capturing console reporter.

And that can be the answer to the "reprinting in report". Just create a `capturing-console` reporter, which will print the captured output at the end of a test case.

In case the stdout/stderr contains an independent diagnostics from the tested codebase, one may want to preserve the stream exactly as-is. We can easily achieve that by adding a capturing reporter that just prints the captured stdout/stderr. So we would have `./test -r xml:out.xml -r 'stdout:&1' -r 'stderr:&2'`, where `&1` and `&2` would be a special path syntax to specify stdout/stderr.

---

**TL;DR**: If at least one capturing reporter is specified, the stdio is captured. If one wishes to include it in a report, they should use an appropriate capturing reporter for that use case. If one wishes to capture the stdout/stderr separately or reprint it to stdout/stderr, they should use special reporters `stdout`/`stderr` which just reprint the captured output to the specified stream.

Tasks:
* [ ] do not reprint anything
* [ ] create a `capturing-console` reporter which will print the captured stdio at the end of each test case
  * to demonstrate the possibility; when per-reporter options are implemented, we
    can make it an option for the console reporter (but that's for later)
* [ ] create `stdout`, `stderr` reporters, which will just print the captured stdout/stderr at the end of each test case
* [ ] add support for `&1`, `&2` output path specifications

Reply:
Ok, so I've thought about this a bit, and I agree in regards to correctness.

I had an idea about having the capture mechanism be transparent in cases where the uncapturing reporter is the one writing to stdout, which could be done (with some mild annoyance) with the capture mechanisms, but then I decided that it seems like too much work.

The reason was that by the time I am down to printf debugging, I am usually kinda desperate and want the output to happen exactly when written and as written, without further surprises. However, then I realized that I don't need to have multiple reporters in these cases 🤷